src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -0,0 +1,477 @@
+        if int_buffer[0] == 0:
+            raise _TimeoutError(
+                "Timeout connecting to streaming request",
+                request=request,
+                response=None,
+            )
+        if int_buffer[0] == SUCCESS_HEADER:
+            # got response
+            # header length is in second int of intBuffer
+            string_len = int_buffer[1]
+            # decode the rest to a JSON string
+            decoder = js.TextDecoder.new()
+            # this does a copy (the slice) because decode can't work on shared array
+            # for some silly reason
+            json_str = decoder.decode(byte_buffer.slice(0, string_len))
+            # get it as an object
+            response_obj = json.loads(json_str)
+            return EmscriptenResponse(
+                request=request,
+                status_code=response_obj["status"],
+                headers=response_obj["headers"],
+                body=io.BufferedReader(
+                    _ReadStream(
+                        int_buffer,
+                        byte_buffer,
+                        request.timeout,
+                        self._worker,
+                        response_obj["connectionID"],
+                    ),
+                    buffer_size=1048576,
+                ),
+            )
+        elif int_buffer[0] == ERROR_EXCEPTION:
+            string_len = int_buffer[1]
+            # decode the error string
+            decoder = js.TextDecoder.new()
+            json_str = decoder.decode(byte_buffer.slice(0, string_len))
+            raise _StreamingError(
+                f"Exception thrown in fetch: {json_str}", request=request, response=None
+            )
+        else:
+            raise _StreamingError(
+                f"Unknown status from worker in fetch: {int_buffer[0]}",
+                request=request,
+                response=None,
+            )
+
+
+# check if we are in a worker or not
+def is_in_browser_main_thread() -> bool:
+    return hasattr(js, "window") and hasattr(js, "self") and js.self == js.window
+
+
+def is_cross_origin_isolated() -> bool:
+    return hasattr(js, "crossOriginIsolated") and js.crossOriginIsolated
+
+
+def is_in_node() -> bool:
+    return (
+        hasattr(js, "process")
+        and hasattr(js.process, "release")
+        and hasattr(js.process.release, "name")
+        and js.process.release.name == "node"
+    )
+
+
+def is_worker_available() -> bool:
+    return hasattr(js, "Worker") and hasattr(js, "Blob")
+
+
+_fetcher: _StreamingFetcher | None = None
+
+if is_worker_available() and (
+    (is_cross_origin_isolated() and not is_in_browser_main_thread())
+    and (not is_in_node())
+):
+    _fetcher = _StreamingFetcher()
+else:
+    _fetcher = None
+
+
+def send_streaming_request(request: EmscriptenRequest) -> EmscriptenResponse | None:
+    if _fetcher and streaming_ready():
+        return _fetcher.send(request)
+    else:
+        _show_streaming_warning()
```

Comment:
Instead of warning about falling back on non-streaming, what do you think about erroring out with a link to documentation about how to properly do streaming w/ web workers? Then users will always get the behavior they are expecting.

The same might be applicable to timeouts? Although we might need people to opt-in to "browser controlled timeouts" via `urllib3.contrib.emscripten.BROWSER_TIMEOUT` sentinel or the like?

Reply:
I think the risk here is that if people use library code that itself uses urllib3 and sets it it to stream by default, they won't be happy if their use case fails in e.g. pyscript or pyodide without web-workers even when it doesn't absolutely need streaming to function. Streaming with web-workers requires cross-site isolation which is a bit fiddly, especially for non-trivial websites, and also doesn't work easily on some big hosting providers including github pages.

I think this is all a transitional thing that will keep things running until the webassembly javascript promise integration (JSPI) proposal comes to all browsers, which allows webassembly code to suspend waiting on a javascript promise, which fixes all this stuff, but until then, I'd rather do whatever breaks least existing code.

( JSPI = https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md )

Reply:
We can always go that route later if we decide it's too much of a hassle, when you set a default behavior it's forever :) From my POV, almost all packages that use urllib3 don't use the streaming functionality and the ones that do either let you control that (like Requests) or are fetching gigantic files likely to be written to the disk (something you wouldn't do in a browser).

If we wait for JSPI we can then offer streaming functionality safely without this behavior. What do you think about this approach?